File: ruby/luhn/luhn.rb
# comment
module Luhn
  module_function

  def luhn_sum(digits)
    digits.reverse.each_with_index.reduce(0) do |sum, (n, i)|
      nn = n * ((i % 2).odd? ? 2 : 1)
      nn -= 9 if nn > 9
      sum + nn
    end
  end

  def valid?(input)
    return false if input =~ /[^[:digit:][:blank:]]/

    digits = input.gsub(/\D/, '').chars.map(&:to_i)
    return false if digits.length < 2

    (luhn_sum(digits) % 10).zero?
  end
end

File: go/luhn/luhn.go
package luhn

var digitValue = [][]int{
	{0, 1, 2, 3, 4, 5, 6, 7, 8, 9},
	{0, 2, 4, 6, 8, 1, 3, 5, 7, 9},
}

func Valid(id string) bool {
	sum := 0
	idx := 0

	for i := len(id) - 1; i >= 0; i-- {
		c := int(id[i])
		switch {
		case c == 32: // ignore spaces
		case 48 <= c && c <= 57: // digit
			sum += digitValue[idx % 2][c - 48]
			idx++
		default:
			return false
		}
	}

	return idx >= 2 && sum % 10 == 0
}
